nifi-nar-bundles/nifi-beats-bundle/nifi-beats-processors/src/main/java/org/apache/nifi/processors/beats/ListenBeats.java

```diff
@@ -0,0 +1,216 @@
+        beatsEncoder = new BeatsEncoder();
+    }
+
+    @Override
+    protected ChannelDispatcher createDispatcher(final ProcessContext context, final BlockingQueue<BeatsEvent> events) throws IOException {
+        final EventFactory<BeatsEvent> eventFactory = new BeatsEventFactory();
+        final ChannelHandlerFactory<BeatsEvent, AsyncChannelDispatcher> handlerFactory = new BeatsSocketChannelHandlerFactory<>();
+
+        final int maxConnections = context.getProperty(MAX_CONNECTIONS).asInteger();
+        final int bufferSize = context.getProperty(RECV_BUFFER_SIZE).asDataSize(DataUnit.B).intValue();
+        final Charset charSet = Charset.forName(context.getProperty(CHARSET).getValue());
+
+        // initialize the buffer pool based on max number of connections and the buffer size
+        final BlockingQueue<ByteBuffer> bufferPool = createBufferPool(maxConnections, bufferSize);
+
+        // if an SSLContextService was provided then create an SSLContext to pass down to the dispatcher
+        SSLContext sslContext = null;
+        final SSLContextService sslContextService = context.getProperty(SSL_CONTEXT_SERVICE).asControllerService(SSLContextService.class);
+        if (sslContextService != null) {
+            sslContext = sslContextService.createSSLContext(SSLContextService.ClientAuth.REQUIRED);
+        }
+
+        // if we decide to support SSL then get the context and pass it in here
+        return new SocketChannelDispatcher<>(eventFactory, handlerFactory, bufferPool, events,
+            getLogger(), maxConnections, sslContext, charSet);
+    }
+
+
+    @Override
+    protected String getBatchKey(BeatsEvent event) {
+        return event.getSender();
+    }
+
+    protected void respond(final BeatsEvent event, final BeatsResponse beatsResponse) {
+        final ChannelResponse response = new BeatsChannelResponse(beatsEncoder, beatsResponse);
+
+        final ChannelResponder responder = event.getResponder();
+        responder.addResponse(response);
+        try {
+            responder.respond();
+        } catch (IOException e) {
+            getLogger().error("Error sending response for transaction {} due to {}",
+                new Object[]{event.getSeqNumber(), e.getMessage()}, e);
+        }
+    }
+
+    protected void postProcess(final ProcessContext context, final ProcessSession session, final List<BeatsEvent> events) {
+        // first commit the session so we guarantee we have all the events successfully
+        // written to FlowFiles and transferred to the success relationship
+        session.commit();
+        // respond to each event to acknowledge successful receipt
+        for (final BeatsEvent event : events) {
+            respond(event, BeatsResponse.ok(event.getSeqNumber()));
+        }
+    }
+
+    @Override
+    protected String getTransitUri(FlowFileEventBatch batch) {
+        final String sender = batch.getEvents().get(0).getSender();
```

Comment:
I would say not:

final InetAddress sender = socketChannel.socket().getInetAddress();

But some people also assumed the Titanic was unsinkable... :ship: